src/dataLoaderService.ts

```diff
@@ -0,0 +1,114 @@
+import DataLoader, { BatchLoadFn } from 'dataloader';
+import { Context } from './Context';
+import { getShortUrl } from './common';
+import { SourceMember } from './entity';
+import { GQLSource } from './schema/sources';
+
+export const defaultCacheKeyFn = <K>(key: K) => {
+  if (typeof key === 'object') {
+    return JSON.stringify(key);
+  }
+
+  return key.toString();
+};
+
+export class DataLoaderService {
+  protected loaders: Record<string, DataLoader<unknown, unknown>>;
+  private ctx: Context;
+
+  constructor({ ctx }: { ctx: Context }) {
+    this.loaders = {};
+    this.ctx = ctx;
+  }
+
+  protected getLoader<K, V>({
+    type,
+    loadFn,
+    cacheKeyFn,
+  }: {
+    type: string;
+    loadFn: (params: K) => Promise<V> | V;
+    cacheKeyFn: (key: K) => string;
```

Comment:
I made `cacheKeyFn` required when creating a new loader so that we can customize the keys per loader.

For example `referralUrl` data loader accepts `source` but if we used the default cache key fn the key would actually contain many properties depending on the `source` object you provided which would then create cache misses because `source` which has `currentMember` and `source` without it would be different cache keys.

By providing custom cache key below we can say that for `referralUrl` we wanna cache per user per source and everything else is irrelevant.